### PR TITLE
fix #40

### LINKED ISSUE #40: SecurityToken's startSTO & endSTO can be set to values different to the actual STO's start/end times
ST. selectOfferingProposal() gets passed startSTO and endSTO parameters which could be different times that those set on an actual STO. 

Either:
1. These get read from the STO contract instead of being set and we eliminate these variables from the ST contract (which IMO they don't belong in the ST contract at all).
2. Or it should be required that the values entered match the values set on the STO.

There's also the issue that there's no way to enforce these two variables stay in sync with the actual STO. As it is, even if we enforce they should match, the owner of the STO could modify the values later on if their contract allows so.

### PR DESCRIPTION
**Pull request by satyamakgec**

### CODE REVIEW

File: contracts/SecurityToken.sol [hunk 1]
```diff
@@ -188,31 +185,24 @@ contract SecurityToken is IERC20 {
...
+    function selectOfferingProposal (uint8 _offeringProposalIndex) public onlyDelegate returns (bool success) {   
         require(!isSTOProposed);
         var (_stoContract, _auditor, _vestingPeriod, _quorum, _fee) = PolyCompliance.getOfferingByProposal(this, _offeringProposalIndex);
         require(_stoContract != address(0));
         require(merkleRoot != 0x0);
         require(delegate != address(0));
-        require(_startTime > now && _endTime > _startTime);
         require(POLY.balanceOf(this) >= allocations[delegate].amount.add(_fee));
+        STO = STO20(_stoContract);
+        require(STO.endTime() > now && STO.endTime() > STO.startTime());
```
Comment (adamdossa) [inline contracts/SecurityToken.sol]: Should this be:
require(STO.startTime() > now && STO.endTime() > STO.startTime());
instead?

Comment (satyamakgec) [reply contracts/SecurityToken.sol]: yep

**Comment by adamdossa** [comment]
Review:

  - In STOContract you don't need to reimplement the `onlyST` method as this is inherited from `STO20` (note `STO20` uses `token` rather than `securityTokenAddress` which should be made consistent between the two files.

  - Currently the logic split between STO and ST is broken. e.g. start/endTime are not used in STO, but are set (by the ST) & `selectOfferingProposal` should be in the STO not the ST. There needs to be a better separation of concerns between these two contracts IMO.

File: contracts/SecurityToken.sol [hunk 2]
```diff
@@ -51,9 +53,8 @@ contract SecurityToken is IERC20 {
     mapping(address => Shareholder) public shareholders;              // Mapping that holds the data of the shareholder corresponding to investor address
 
     // STO 
-    address public STO;                                               // Address of the security token offering contract
-    uint256 public maxPoly;                                           // Maximum amount of POLY will be invested in offering contract
     bool public isSTOProposed = false;
+    bool public isOfferingStart = false;
```
Comment (pabloruiz55) [inline contracts/SecurityToken.sol]: Rename to hasOfferingStarted

File: contracts/SecurityToken.sol [hunk 3]
```diff
@@ -180,39 +179,48 @@ contract SecurityToken is IERC20 {
...
+     * @dev Start the offering by sending all the tokens to STO contract
+     * @return bool
+     */
+    function startOffering() onlyOwner external returns (bool success) {
+        require(isSTOProposed);
+        require(!isOfferingStart);
+        uint256 tokenAmount = this.balanceOf(msg.sender);
+        require(tokenAmount == totalSupply);
+        balances[STO] = balances[STO].add(tokenAmount);
+        balances[msg.sender] = balances[msg.sender].sub(balances[STO]);
```
Comment (pabloruiz55) [inline contracts/SecurityToken.sol]: Should be: balances[msg.sender] = balances[msg.sender].sub(tokenAmount);

Comment (satyamakgec) [reply contracts/SecurityToken.sol]: done 0789a27

**Comment by adamdossa** [comment]
Any reason `decimals` is hard-coded to 0? Is the intention that all ST are non-divisible?

**Comment by pabloruiz55** [comment]
> Any reason decimals is hard-coded to 0? Is the intention that all ST are non-divisible?

@adamdossa this was discussed on #20. I'm still on the fence about that. My gut tells me it shouldn't be hardcoded to 0, but it's true I haven't been able to find a reason why shares would be divisible.

**Comment by adamdossa** [comment]
Main reason I think would be liquidity. i.e. an organisation issues an ST, at an initial price of $1. After doing very well, the token price is now $100. Without the token being divisible it wouldn't be possible to invest an amount under $100, which may hurt the market. In equities stocks are often split in these circumstances, and I feel the equivalent in the blockchain world is divisible tokens.

**Comment by adamdossa** [comment]
I've reviewed the code, and my only comment is that we should be consistent on how we use `transfer` and `transferFrom`. Both functions return a boolean and should generally be wrapped in a `require` by the calling function, but this is not done consistently throughout the code. In general this works fine as these functions tend to internally throw / revert on errors, but we should be consistent as the ERC20 interface doesn't guarantee this.

**Comment by pabloruiz55** [comment]
> Main reason I think would be liquidity....

Let's move this conversation to #20

**Comment by pabloruiz55** [comment]
> I've reviewed the code, and my only comment is that we should be consistent on how we use transfer and transferFrom. Both functions return a boolean and should generally be wrapped in a require by the calling function, but this is not done consistently throughout the code. In general this works fine as these functions tend to internally throw / revert on errors, but we should be consistent as the ERC20 interface doesn't guarantee this.

@satyamakgec could you review this plus my 2 small comments and merge before you call it a day, so we can keep working today on the latest?

**Review by pabloruiz55 — APPROVED** [review]

**Pull request merged**